File: backend/dataset/numIslands/numislands_13.py
from typing import List
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:

        island_count = 0

        for i, _ in enumerate(grid):
            for j, _ in enumerate(grid[i]):
                if grid[i][j] == '1':
                    self.dfs(i, j, grid)
                    island_count += 1

        return island_count


    def dfs(self, i, j, grid):

        if i < 0 or i >= len(grid):
            return

        if j < 0 or j >= len(grid[0]):
            return

        if grid[i][j] == '0':
            return

        grid[i][j] = '0'

        self.dfs(i - 1, j, grid)
        self.dfs(i + 1, j, grid)
        self.dfs(i, j - 1, grid)
        self.dfs(i, j + 1, grid)

File: backend/CDataset/numIslands/numIslands_13.cpp
class Solution {
public:
    void dfs(vector<vector<char>>& grid,int i,int j,int r,int c)
    {   if(i<0||j<0||i>=r||j>=c||grid[i][j]!='1') return;
        grid[i][j]='2';
        dfs(grid,i-1,j,r,c);
        dfs(grid,i+1,j,r,c);
        dfs(grid,i,j-1,r,c);
        dfs(grid,i,j+1,r,c);
    }
    int numIslands(vector<vector<char>>& grid) {
        int r=grid.size();
        if(grid.size()==0)
            return 0;
        int c=grid[0].size();
        int ans=0;
        for(int i=0;i<r;i++)
        {  for(int j=0;j<c;j++)
            {   if(grid[i][j]=='1')
                {   dfs(grid,i,j,r,c);
                    ans++;
                }
            }
        }
        return ans;
    }
};
